matching-service/controller/match-controller.js

```diff
@@ -1,106 +1,203 @@
 import {
-  ormCreatePendingMatch as _createPendingMatch,
-  ormGetFirstMatch as _getFirstMatch,
-  ormDeletePendingMatch as _deletePendingMatch,
+    ormCreatePendingMatch as _createPendingMatch,
+    ormDeletePendingMatch as _deletePendingMatch,
+    ormDeleteIfPedningMatchExists as _deleteIfPendingMatchExists,
+    ormGetMatchWithDifficulty as _getMatchWithDifficulty,
+    ormGetMatchWithTopic as _getMatchWithTopic,
 } from "../model/pending-match-orm.js";
 
+import { ormCreateMatchInfo as _createMatchInfo } from "../model/match-info-orm.js";
+
 // Expressed in milliseconds
 export const TIMEOUT_INTERVAL = 5000;
 
 /**
  * Handles the logic of matching users.
  */
 export const matchController = (io, socket) => {
-  console.log(`IO: Socket with id: ${socket.id} connected`);
-
-  socket.on("findMatch", async ({ username, difficultyLevel }) => {
-    const { hasFoundMatch, partnerUsername, partnerSocketId } = await findMatch(
-      difficultyLevel
-    );
-
-    if (hasFoundMatch) {
-      await handleMatchSuccess(
-        io,
-        socket,
-        username,
-        partnerUsername,
-        partnerSocketId
-      );
-      return;
-    }
-    await createPendingMatch(username, socket.id, difficultyLevel);
-    handleMatchFail(io, socket, username);
-  });
+    console.log(`IO: Socket with id: ${socket.id} connected`);
+
+    socket.on("findMatch", async ({ username, difficultyLevel, topic }) => {
+        if (topic === undefined) {
+            topic = null;
+        }
+
+        if (difficultyLevel === undefined) {
+            difficultyLevel = null;
+        }
+
+        if (topic == null && difficultyLevel == null) {
+            await handleInvalidMatchFail(
+                io,
+                socket,
+                "At least difficulty Level of topic level must exist"
```

Comment:
#nit: I think there's a slight typo in your message, should be "or" instead of "of"

"At least difficulty level or topic must exist"